db/commitlog/commitlog.cc

```diff
@@ -2050,10 +2050,9 @@ future<db::rp_handle> db::commitlog::add(const cf_id_type& id,
             res = std::move(h);
         }
     };
-    auto writer = ::make_shared<serializer_func_entry_writer>(id, size, std::move(func), sync);
-    return _segment_manager->allocate_when_possible(*writer, timeout).then([writer] {
-        return std::move(writer->res);
-    });
+    serializer_func_entry_writer writer(id, size, std::move(func), sync);
+    co_await _segment_manager->allocate_when_possible(writer, timeout);
+    co_return std::move(writer.res);
```

Comment:
Again a fast-path allocation.

Reply:
forceinline?

Reply:
Won't work, AFAIK. The compiler doesn't have the smarts to figure out that one coroutine's lifetime is fully enclosed in the other's.

Reply:
Seems odd, feels like an obvious optimization, and a direct result of inline (assuming done before translating coroutine operators to code), but should be easy enough to check with some example code and disassembly...

Reply:
perf_simple_query will also show it as an extra allocation.

It's not a trivial optimization. In fact I'm not sure the compiler is allowed to do it.

Consider

```c++
future<> f() {
    co_await [] () -> future<> {
         co_return g();
    }();
}
```

We can easily see the lambda is contained in f(), but from the compiler's point of view, the completion of the lambda calls lambda::_promise::set_value() or lambda::_promise::set_exception() and does not directly resume f. As far as the compiler knows we might resume f earlier and leave the lambda dangling.

Reply:
Given that the lambda is not a suspension point, it seems to me that (again assuming inline before expanding coroutines proper), the lambda can be inlined just fine, and just turns into 
```
promise p;
try {
   p = g();
} catch (...) [
  g = curr_ex;
} 
co_await p;
```
But again, I am not sure at which stage what happens in clang. As soon as I've benchmarked, I can test some examples. :-)

Reply:
Removed superfluous coroutine frames manually.